paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java

```diff
@@ -404,8 +405,23 @@ private static boolean isPaimonTable(Table table) {
     @Override
     public TableSchema getDataTableSchema(Identifier identifier, String branchName)
             throws TableNotExistException {
-        assertMainBranch(branchName);
-        return getDataTableSchema(identifier);
+        if (!tableExists(identifier)) {
+            throw new TableNotExistException(identifier);
+        }
+        if (branchName.equals(DEFAULT_MAIN_BRANCH)) {
+            return getDataTableSchema(identifier);
+        } else {
+            Path tableLocation = getDataTableLocation(identifier);
+            return new SchemaManager(fileIO, tableLocation, branchName)
+                    .latest()
+                    .map(
+                            s -> {
+                                Options branchOptions = new Options(s.options());
+                                branchOptions.set(CoreOptions.BRANCH, branchName);
+                                return s.copy(branchOptions.toMap());
+                            })
+                    .orElseThrow(() -> new TableNotExistException(identifier));
```

Comment:
Move this into `tableSchemaInFileSystem`, so that it supports reading schema for branches. `FileSystemCatalog` can also reuse this method.